projects/cm_mcu/MonitorI2CTask.c

```diff
@@ -0,0 +1,195 @@
+#include "common/smbus_helper.h"
+#include "common/smbus_units.h"
+#include "MonitorI2CTask.h"
+#include "common/power_ctl.h"
+#include "common/log.h"
+#include "Tasks.h"
+#include "I2CCommunication.h"
+
+// local prototype
+
+void Print(const char *str);
+
+// break out of loop, releasing semaphore if we have it
+
+#define release_break()           \
+  {                               \
+    xSemaphoreGive(args->xSem);   \
+    break;                        \
+  }
+
+// get a mask of a register address given a starting bit field (shift) and the width
+
+void make_bitmask(unsigned int width, unsigned int shift, uint8_t *mask)
+{
+  unsigned int t = width ? (2u << (width - 1)) - 1u : 0u;
+  *mask = t << shift;
+}
+
+// read-only accessor functions for Firefly names and values.
+
+bool getFFch_low(uint8_t val, int channel)
+{
+  configASSERT(channel < 8);
+  if (!((1 << channel) & val)) {
+    return false;
+  }
+  return true;
+}
+
+bool getFFch_high(uint8_t val, int channel)
+{
+  configASSERT(channel >= 8);
+  if (!((1 << (channel - 8)) & val)) {
+    return false;
+  }
+  return true;
+}
+
+extern struct zynqmon_data_t zynqmon_data[ZM_NUM_ENTRIES];
+
+#define TMPBUFFER_SZ 96
+
+// Monitor registers of FF temperatures, voltages, currents, and ClK statuses via I2C
+void MonitorI2CTask(void *parameters)
+{
+
+  struct MonitorI2CTaskArgs_t *args = parameters;
+
+  configASSERT(args->name != 0);
+
+  // watchdog info
+  task_watchdog_register_task(kWatchdogTaskID_MonitorI2CTask);
+
+  // initialize to the current tick time
+  args->updateTick = xTaskGetTickCount();
+
+  // wait for the power to come up
+  vTaskDelayUntil(&(args->updateTick), pdMS_TO_TICKS(2500));
+
+  int IsCLK = (strstr(args->name, "CLK") != NULL);   // the instance is of CLK-device type
+  int IsFF12 = (strstr(args->name, "FF12") != NULL); // the instance is of FF 12-ch part type
+  // int IsFFDAQ =  (strstr(args->name, "FFDAQ") != NULL);  //the instance is of FF 4-ch part type (DAQ links) -- not being used currently
+
+
+  // reset the wake time to account for the time spent in any work in i2c tasks
+
+  bool good = false;
+  for (;;) {
+    char tmp[TMPBUFFER_SZ];
+
+    // grab the semaphore to ensure unique access to I2C controller
+    while (xSemaphoreTake(args->xSem, (TickType_t)10) == pdFALSE)
+        ;
+    args->updateTick = xTaskGetTickCount(); // current time in ticks
+    // -------------------------------
+    // loop over devices in the device-type instance
+    // -------------------------------
+    for (int ps = 0; ps < args->n_devices; ++ps) {
+
+      if (!IsCLK) { // Fireflies need to be checked if the links are connected or not
+        int offsetFF12 = 1 - IsFF12;
+        if (!isEnabledFF(ps + (offsetFF12 * (NFIREFLIES_IT_F1)) + ((args->i2c_dev - I2C_DEVICE_F1) * (-1) * (NFIREFLIES_F2)))) // skip the FF if it's not enabled via the FF config
+          continue;
+        args->updateTick = xTaskGetTickCount();
+      }
+
+      if (args->requirePower) {
```

Comment:
I think all these devices require power, so we can just always institute this check, and get rid of the argument in the `struct MonitorI2CTaskArgs_t` and the check for it here.

Reply:
Should I do the same thing with MonTask? because I can the power check here: https://github.com/apollo-lhc/cm_mcu/blob/new_monitor_i2c_task/projects/cm_mcu/MonitorTask.c#L83

Reply:
no, there the power supplies and FPGAs are treated differently